internal/uvm/cpugroups.go

```diff
@@ -0,0 +1,158 @@
+package uvm
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const cpuGroupNullID = "00000000-0000-0000-0000-000000000000"
+
+var HV_STATUS_INVALID_CPU_GROUP_STATE = errors.New("The hypervisor could not perform the operation because the CPU group is entering or in an invalid state.")
+
+// ReleaseCPUGroup unsets the cpugroup from the VM and attemps to delete it
+func (uvm *UtilityVM) ReleaseCPUGroup(ctx context.Context) error {
+	groupID := uvm.cpuGroupID
+	if groupID == "" {
+		// not set, don't try to do anything
+		return nil
+	}
+	if err := uvm.unsetCPUGroup(ctx); err != nil {
+		return fmt.Errorf("failed to remove VM %s from cpugroup %s", uvm.ID(), groupID)
+	}
+
+	err := DeleteCPUGroup(ctx, groupID)
+	if err != nil && err == HV_STATUS_INVALID_CPU_GROUP_STATE {
+		log.G(ctx).WithField("error", err).Warn("cpugroup could not be deleted, other VMs may be in this group")
+		return nil
+	}
+	return err
+}
+
+// ConfigureVMCPUGroup setups up the cpugroup for the VM with the requested id
+func (uvm *UtilityVM) ConfigureVMCPUGroup(ctx context.Context, id string) error {
+	if id == "" {
+		return fmt.Errorf("must specify an ID to use when configuring a VM's cpugroup")
+	}
+	exists, err := cpuGroupExists(ctx, id)
```

Comment:
What error does the modify call give if the group doesn't exist? I'm wondering if we couldn't just skip the existence check, and rely on HCS to give us a relevant error in that case.

Reply:
I believe it will be an "element not found" error message. I personally find that descriptive enough, what do you think?

Reply:
Good enough for me!

Reply:
I took it out :)